src/vos/vos_io.c

```diff
@@ -1561,31 +1563,62 @@ dkey_update_begin(struct vos_io_context *ioc)
 }
 
 int
-vos_publish_scm(struct vos_container *cont, struct vos_rsrvd_scm *rsrvd_scm,
-		bool publish)
+vos_publish_scm(struct vos_container *cont, struct vos_rsrvd_scm **rsrvd_scm,
+		bool publish, struct dtx_handle *dth)
 {
 	int	rc = 0;
 
-	D_ASSERT(rsrvd_scm->rs_actv_at <= rsrvd_scm->rs_actv_cnt);
-	if (rsrvd_scm->rs_actv_at == 0)
+	if (*rsrvd_scm == NULL)
 		return 0;
 
-	D_ASSERT(rsrvd_scm->rs_actv != NULL);
-	if (publish)
-		rc = umem_tx_publish(vos_cont2umm(cont), rsrvd_scm->rs_actv,
-				     rsrvd_scm->rs_actv_at);
-	else
-		umem_cancel(vos_cont2umm(cont), rsrvd_scm->rs_actv,
-			    rsrvd_scm->rs_actv_at);
+	if (publish && (*rsrvd_scm)->rs_published)
+		return 0;
+
+	if (!publish && (*rsrvd_scm)->rs_cancelled)
+		return 0;
+
+	D_ASSERT((*rsrvd_scm)->rs_actv_at <= (*rsrvd_scm)->rs_actv_cnt);
+
+	if (publish) {
+		if (dth != NULL) {
+			struct dtx_rsrvd_uint	*dru;
+
+			D_ASSERT(dth->dth_op_seq >= 1);
+
+			dru = &dth->dth_rsrvds[dth->dth_op_seq - 1];
+			dru->dru_scm_ptr = *rsrvd_scm;
+			D_INIT_LIST_HEAD(&dru->dru_nvme_list);
+
+			/* @rsrvd_scm is taken over. */
+			*rsrvd_scm = NULL;
+
+			return 0;
+		}
+
+		rc = umem_tx_publish(vos_cont2umm(cont), (*rsrvd_scm)->rs_actv,
+				     (*rsrvd_scm)->rs_actv_at);
+	} else {
+		umem_cancel(vos_cont2umm(cont), (*rsrvd_scm)->rs_actv,
+			    (*rsrvd_scm)->rs_actv_at);
+	}
+
+	if (rc != 0) {
+		D_ERROR("%s SCM reservations failed: "DF_RC"\n",
+			publish ? "Publish" : "Cancel", DP_RC(rc));
+	} else {
+		if (publish)
+			(*rsrvd_scm)->rs_published = 1;
+		else
+			(*rsrvd_scm)->rs_cancelled = 1;
```

Comment:
Originally, they were for aggregation case because the 'vos_rsrvd_scm' space may be reused by aggregation. I will check what will happen after removing it.